torch/utils/cpp_extension.py

```diff
@@ -0,0 +1,188 @@
+            root_extensions_directory))
+
+    build_directory = os.path.join(root_extensions_directory, name)
+    if not os.path.exists(build_directory):
+        if verbose:
+            print('Creating extension directory {}...'.format(build_directory))
+        # This is like mkdir -p, i.e. will also create parent directories.
+        os.makedirs(build_directory)
+
+    return build_directory
+
+
+def _build_extension_module(name, build_directory):
+    try:
+        subprocess.check_output(
+            ['ninja', '-v'], stderr=subprocess.STDOUT, cwd=build_directory)
+    except subprocess.CalledProcessError:
+        # Python 2 and 3 compatible way of getting the error object.
+        _, error, _ = sys.exc_info()
+        # error.output contains the stdout and stderr of the build attempt.
+        raise RuntimeError("Error building extension '{}': {}".format(
+            name, error.output.decode()))
+
+
+def _import_module_from_library(module_name, path):
+    # https://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path
+    file, path, description = imp.find_module(module_name, [path])
+    # Close the .so file after load.
+    with file:
+        return imp.load_module(module_name, file, path, description)
+
+
+def _write_ninja_file(
+        path,
+        name,
+        sources,
+        extra_cflags,
+        extra_ldflags,
+        extra_include_paths):
+    try:
+        import ninja
+    except ImportError:
+        raise RuntimeError("Ninja is required to load C++ extensions. "
+                           "Install it with 'pip install ninja'.")
+    with open(path, 'w') as build_file:
+        writer = ninja.Writer(build_file)
+        writer.variable('ninja_required_version', '1.3')
+        writer.variable('cxx', os.environ.get('CXX', 'c++'))
+
+        # Turn into absolute paths so we can emit them into the ninja build
+        # file wherever it is.
+        sources = [os.path.abspath(file) for file in sources]
+        includes = [os.path.abspath(file) for file in extra_include_paths]
+
+        # include_paths() gives us the location of torch/torch.h
+        includes += include_paths()
+        # sysconfig.get_paths()['include'] gives us the location of Python.h
+        includes.append(sysconfig.get_paths()['include'])
+
+        cflags = ['-fPIC', '-std=c++11']
+        cflags += ['-I{}'.format(include) for include in includes]
+        cflags += extra_cflags
+        writer.variable('cflags', ' '.join(cflags))
+
+        ldflags = ['-shared'] + extra_ldflags
+        if sys.platform == 'darwin':
+            ldflags.append('-undefined dynamic_lookup')
+        writer.variable('ldflags', ' '.join(ldflags))
+
+        writer.rule(
+            'compile',
+            command='$cxx -MMD -MF $out.d $ldflags $cflags $in -o $out',
+            depfile='$out.d',
+            deps='gcc')
+
+        target = '{}.so'.format(name)
+        writer.build(outputs=target, rule='compile', inputs=sources)
```

Comment:
We should have per-file builds and a separate link (`g++ foo.cpp -o foo.o`, `g++ foo.o bar.o`) -- this will allow incremental compilation and parallel builds. This can be in a follow up PR if you want.